### ISSUE TITLE
gc + shared_ptr causes segfault

### ISSUE THREAD

**Issue by JMLX42** [Tomasu/LuaGlue#15]
Hi there,

I've slightly modified you shared_ptr example:

<pre>
-- shared_ptr.lua
function test_ptr(ptr)
    print("test_ptr")
    print(ptr)
end
</pre>

<pre>
// shared_ptr.cpp
state.invokeVoidFunction("test_ptr", ptr);
state.invokeVoidFunction("test_ptr", ptr);
state.invokeVoidFunction("test_ptr", ptr);
</pre>

compile/run gives a segfault with CC 4.8.1. It's triggered by Lua's GC that gets exectuted as the LuaGlue object gets destroyed. I haven't tried to do the same kind of things without shared_ptr, but I suspect the two are closely related.

Any idea?

**Comment by JMLX42** [comment]
I think the problem comes from:

https://github.com/Tomasu/LuaGlue/blob/master/include/LuaGlue/LuaGlueApplyTuple.h#L138

<pre>
std::shared_ptr<T> *ptr = new std::shared_ptr<T>(v);
</pre>

Creating a ne shared_ptr object should never be done because it will create an alternate reference counter. When this reference counter reaches 0, then the wrapped object will be destroyed despite the fact that it is still referenced by the original shared_ptr.

In other words: you should never do a new std::shared_ptr on an object you got from a shared_ptr.

How could we fix this?

**Comment by Tomasu** [comment]
I think you can create a new shared_ptr that references an old one... but yes that's a problem. Temp fix, never send in un-bound types. The only time that code is executed is when it's passed a type that isn't registered.

If you need to pass in unregistered types, then something else is needed, but I'm not sure what, because something needs to be stored in that lua type, but we can't just pass in a pointer to the shared_ptr we're passed, as its a temporary. I think we could just wrap it in a LuaGlueObject directly, but if its taken from a shared_ptr, we "get" a reference to the shared_ptr so it isn't freed prematurely. that /may/ work.

**Comment by JMLX42** [comment]
I didn't meant you "cannot" I meant "you should definitely not" :)
So any class bound using LuaGlue.Class() will _NOT_ use this code?

**Comment by Tomasu** [comment]
Ah, or a special implementation of LuaGlueObject that stores a shared_ptr<T> rather than a Type*. that may work better, and let us take advantage of shared_ptr's own internal ref counting.

**Comment by Tomasu** [comment]
Yeah, bound classes should never ever hit the code at the bottom of that function. It's entirely to pass around objects it doesn't know how to handle.

**Comment by JMLX42** [comment]
Still there is this segfault issue...

**Comment by Tomasu** [comment]
Yeah, I'll have to take a look at it soon. I had wanted to on the weekend, but stuff came up.

**Comment by Tomasu** [comment]
It turns out that it wasn't anything to do with shared_ptrs or that you can't new shared_ptr's (you can, just make sure you use the copy constructor, or operator= to get things to work properly). I had a misplaced lua_pop in the invokeVoidFunction method. It was corrupting the lua stack ever so slowly. 1 call didn't cause a crash, 2 also didn't, but as soon as three calls happened, lua and malloc exploded.

One thing I tried to do before realizing that was converting the LuaGlueObject<shared_ptr> type to only hold onto the shared_ptr.. but that can't work with the current design. the LuaGlueObject's NEED to be able to share "Impl" objects in order to not leak (or cause potential double frees).

**Comment by JMLX42** [comment]
Thanks! I'll check if it fixes my oother corrupted pointer error and let you know.
